README.md
# magical-state

![plot](./assets/logo.png)

## Table of contents

- [Introduction](#introduction)
- [Installation](#installation)
- [Usage](#usage)
  - [Specification](#define-a-specification-in-json-format)
  - [Implementation of fetching data](#add-the-implementation-of-fetching-data)
  - [Implementation of default values](#add-the-implementation-of-getting-default-values)
- [Store methods](#store-methods)
  - [Create the store](#create-the-store)
  - [Change the state of the store](#change-the-state-of-the-store)
  - [Get the state of the store](#get-the-state-of-the-store)
  - [Import / Export Store URL](#import-and-export-store-url)
- [Store Events](#store-events)
- [Independent Methods](#independent-methods)
- [Vue Components](/src/components/vue2-components/)
- [Examples](/examples)

## Introduction

This library manages the global state of a series of selectors which have reactive dependencies between each other. These dependencies and the selectors' expected behaviour can be easily defined through a configuration file.

## Installation

Add the dependency for your proyect into `package.json`

```json
"magical-state": "git+https://gitlab.lbd.org.es/publico/magical-state.git"
```

## Usage

To use the library we must create a [store](#create-the-store). For this we will first need to define a specification and an object consisted of two methods, _getValues_ and _defaultValuesGetter_, that will retrieve the items that will populate each selector and specify a default value to be set.

### Define a `specification` in json format

To begin with, we must define in a json the selectors we want to create, the way they are related to each other, and their expected behaviour. The store will accept the next parameters in the specification:

| name              | type    | default   | description                                                                                                                       |
| ----------------- | ------- | --------- | --------------------------------------------------------------------------------------------------------------------------------- |
| id                | string  | undefined | Identifier of the selector, must be unique.                                                                                       |
| label             | string  | undefined | Label to be placed in the label field of the selector.                                                                            |
| setItemsOnMounted | boolean | false     | It can be specified if we want the selector to load data when it is rendered for the first time.                                  |
| triggerCallback   | boolean | false     | If this option is selected the callback defined passed in store creation will be fired whenever the value of the selector changes |
| actions           | array   | []        | List of identifiers of store elements that need to reload their items every time the value of this element changes.               |

<details>
  <summary>Example of a specification.json</summary>

#### Specification

This spec will generate two selectors: "COUNTRIES" and "CITIES".

_COUNTRIES_ will trigger the _getValues_ method on mount and set as its selector options to the returned elements, setting as its selected value the element retrieved by _defaultValuesGetter_. Everytime the selected value changes _CITIES_ will update its items and, again, _defaultValuesGetter_ will be invoked.

```json
[
  {
    "id": "COUNTRIES",
    "label": "Countries",
    "triggerCallback": true,
    "actions": ["SPATIAL_FILTER"]
  },
  {
    "id": "CITIES",
    "label": "Cities",
    "actions": []
  }
]
```

</details>

##### **IMPORTANT**

Be careful with defining cyclic dependencies between selectors, and that the identifiers in the actions are defined in the specification.

### Add the implementation of fetching data

We must define how we want to retrieve the necessary information to populate each of our selectors.

To do this the store will call the method _getValues_ passed on store creation. This method is expected to receive the next parameters in the listed order:

- _propId_: The identifier of the selector that needs its items to be loaded.
- _params_: A list of key-value elements that contains the current value of the selectors that have dependent selectors (_actions_ list not empty).
- _store_: A key-value object containing all the selectors' values.

<details>
  <summary>Example of a getValues function</summary>

```js
export default async (propId, params, store) => {
  switch (propId) {
    case "COUNTRIES":
      return countriesService.getCountries();
    case "CITIES":
      return citiesService.getCities(params["COUNTRIES"]);
  }
};
```

</details>

##### **IMPORTANT FOR RETURNING VALUES**

The function always must return a Promise, and the retrieved data accepts the next format:

```json
[
  {
    "label": "any",
    "value": "any"
  }
]
```

Field _label_ is only mandatory when using the library's Vue components.

### Add the implementation of getting default values

We must define a method that indicates the store the default values to be set on selectors that are dependent of others (are present in any selectors' actions) or those that are meant to be populated on store creation.

To do this the store will call the method _defaultValuesGetter_ passed on store creation. This method is expected to receive the next parameters in the listed order:

- _propId_: The identifier of the selector that needs its items to be loaded.
- _items_: A list containing the selector's items.
- _store_: A key-value object containing all the selectors' values.

##### **IMPORTANT**

The _defaultValuesGetter_ function is expected to retrieve a Promise containing the desired new value.

<details>
  <summary>Example of a defaultValuesGetter function</summary>

```js
export default async (propId, items, store) => {
  return new Promise((resolve) => {
    switch (propId) {
      // Will set as value the last item
      case "TEMPROAL_FILTER":
        resolve(items[items.length - 1].value);
        break;
      case "DATE_FILTER":
        resolve(formatDateInverse(new Date()));
        break;
      default:
        resolve();
        break;
    }
  });
};
```

</details>

## Store methods

### Create the store

> `createStore (jsonSpec: Array, getValues: Function, state: Object|String, callback: Function) : Store`

- _jsonSpec_: the [specification](#define-a-specification-in-json-format) needed to create the desired selectors.

- {_getValues_: ()=>{}, _defaultValuesGetter_: ()=>{} }: the [fetch](#add-the-implementation-of-fetching-data) data and default values getter functions.

- _state_: a state to be set when the store is created (url encoded or an object with key/values). When passed the store will ignore the specification for value getting/setting and will set the selectors' items based on the values passed.

- _callback_: a _Promise_ to be executed when the store values are updated. The callback function will be triggered everytime a selector that has the _triggerCallback_ property to true suffers a value change. Note that if the selector has any related selectors on its actions, the store will wait for their items and default values to be set before triggering the callback.

- _Returns_ an instance of the store.

<details>
<summary>Example of store creation</summary>

```js
import jsonSpec from "./specification.json";
import {
  createStore
} from "./magical-state/index.js";
import getValues from "./valueGetters";
import defaultValuesGetter from "./defaultValuesGetter";

...
...
...

const state = {
  COUNTRIES: 2,
};

this.store = createStore(
  jsonSpec,
  {
    getValues: getValues,
    defaultValuesGetter: defaultValuesGetter
  },
  // will set 2 as default value on selector 'countries'
  state,
  // the callback receives as a parameter an {id: value} object
  // that constains the selectors with values not null
  (storeCurrentState) => {
    return new Promise(async (resolve) => {
      resolve();
    });
  };
);


```

</details>

### Change the state of the store

- `change(propId: String, newVal: Any, needsRedraw: Boolean): Promise<>`: Changes the value of the selector with the given id. _needsRedraw_ prop defaults to true and specifies if the change on the selector's value should trigger the callback, in case the selector has the _triggerCallback_ property set to true in the specification.
- `setSelector(selectorId: String, newVal: Any, deep: Boolean, triggerCallbak: Boolean): Promise<>`: Changes the value of the selector with the given id. If _deep_ is true, or the selector doesn't have items, _setSelector_ will call _getValues_ to populate the selector's items.
- `setItems(id: String, values: Array, useSpecConfig: Boolean): void`: Sets the items of the specified store element to the parameter _values_.
- `setHasErrors(selectorId: String, value: Boolean, useSpecConfig: Boolean): Boolean`: Sets the property 'hasErrors' of the specified selector to the value provided. If it's set to true this will prevent the store from firing the callback when an element changes value.

##### Important when setting a new value on a selector:

- If a selector has an error (_hasError_ property equaling true) and its value changes to a correct one, it's important to first set this property back to false so that the callback is triggered when calling _change_ function.

### Get the state of the store

- `getSelector(selectorId: String): Proxy`: Returns the proxy of the selector with the given id).
- `(getter) objFromObservable(): Array<{id: String, value: Any}>`: Returns a copy of the actual state of the store.
- `getUI(): Array<{id: String, label: String, value: Any, type: String, items: []}>`: Returns a list of objects that contains the basic information about every observable element so they can be displayed on the UI.
- `triggerGetValues(id: String): Promise<Any>`: Triggers the _getValues_ method of the selector with the given id. Returns the promise returned by the implementation of the _getVlues_ method passed on store creation.

### Import and Export Store URL

For importing and exporting the store, you can use the methods:

- `exportStoreEncodedURL(): String`: Returns an encoded URL with the current state of the store.
- `importStoreEncodedURL(url): void`: Parses the URL and sets the state of the store).
- `(independent from store) parseUrl(url,jsonSpec): Object<key,value>`: Returns the decoded URL, this method can be called without instantiating the store, to parse an URL before store creation.

<details>
<summary>Example of url import/export</summary>

```js
import {
  createStore, parseUrl
} from "./magical-state/index";
import getValues from "./getValues";
import defaultValuesGetter from "./defaultValuesGetter";
...

// If you need to access the readable values of an encoded
//URL, you can use the method:
const newState = parseUrl(url, jsonSpec);

// You can create the new store with an url encoded or
// with an object with key/values
const newStateUrl = "MD0yJjI9Mg=="
const newStateObject = {
  "SPATIAL_AGGREGATION": "PROVINCE",
  "SPATIAL_FILTER": "A CORUÑA",
  "TEMPORAL_AGGREGATION": "YEAR",
  "TEMPORAL_FILTER": "2018"
};

// Examples creating the store
this.store = createStore(
  jsonSpec,
  {getValues: getValues, defaultValuesGetter: defaultValuesGetter},
  newStateUrl || newStateObject,
  (storeContent) => {
    console.log("created with new state");
  }
);

// Then you can import/export the store
this.store.exportStoreEncodedURL()
// returns the encoded URL (MD0yJjI9Mg==)

// set the state of the store with the encoded URL
this.store.importStoreEncodedURL(newStateUrl);

```

</details>

## Store Events

The store will dispatch the next events:

- _redrawFulfilled_: It is dispatched after completing the execution of the callback function.

- _itemsLoaded_: It is dispatched after retrieving the items of a selector through the use of the _getValues_ function.

- _change_: Dispatched when a selector has changed its value.

**Important**: To dispatch these events we use [CustomEvent](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent) so we can append on them iformation about the store. With each of them an object with the next content is passed:

```js
{event:
  {detail:
    {
      id: String,
      label: String,
      value: Any,
      type: String,
      items: Array
    }
  }
}
```

## Independent Methods

`parseUrl(url,jsonSpec): Object<key,value>` Returns the decoded URL.

## Vue2 Components

The library provides some vue2 components that work against a store. You cand find how to use them at: [vue2-components](/src/components/vue2-components/).

### :pencil: Examples

Extended examples at: [Examples](/examples).

## Changelog

[CHANGELOG.md](https://gitlab.lbd.org.es/proyectos-sig/magical-state/-/blob/main/CHANGELOG)

In order to automate CHANGELOG update, developers should write commit messages following [Angular Commit Message Conventions](https://semantic-release.gitbook.io/semantic-release/#commit-message-format)

## License

[MIT](http://opensource.org/licenses/MIT)


CHANGELOG.md
# CHANGELOG

<!--- next entry here -->

## 0.28.1
2023-04-13

### Fixes

- bug in TimelineComponent (d12f9d5bae181d17af789a8c1b907ec1953c10fd)

## 0.28.0

2023-03-14

### Features

- Added defaultValuesGetter option to store (64131af40bb145634c4c7b08db4089952d90181f)

## 0.27.2

2023-02-28

### Fixes

- Added new props and emits to vue2-components README (001e0b512e2c30828e81860c52ddb29631ff535b)
- Added disabled prop to slider component and fixed bug with isDisabled computed (8bf026cd00574d7b75c503563137ca90e37d4813)

## 0.27.1

2022-11-11

### Fixes

- Changed style of timeline component (7f27cd59380adae90b08df089b8a4082d40debc3)
- Added limit buttons prop (02f4aab382ecdd352e141c8d3e57e964dd2d2d97)

## 0.27.0

2022-11-10

### Features

- Added hideDetails prop on vue2-components (23d50b3120b836d478d8b2b8d0f14bf5fc9a1870)

## 0.26.2

2022-11-10

### Fixes

- Fixed not setting previous selected value when rule applies again on another depending selector (e6608daff6ed939977d8358ae2d15ce262369fd0)

## 0.26.1

2022-11-09

### Fixes

- Fixed bug with mAucomplete templates (46aee821b200ebca6d674076d1a1d5c3d8996c72)

## 0.26.0

2022-11-08

### Features

- Added support for modifying the selection slot in mAutoComplete (8ad726557ebe73b1a1906c6f61e6ef93be27cb82)

## 0.25.1

2022-10-26

### Fixes

- **datefiltercomponent:** Added 'firstDayOfWeek' prop (3b52d996df3c626d6117e43b73a4b558902d0e1b)
- Fixed bug with labels (220cc892309f6cf0e4ba67d613e55fd045693374)

## 0.25.0

2022-10-24

### Features

- Changes on rules so that they are completely managed from the store (5655adaec2f4b8f46be9486dccdbeb44dde3bf59)

## 0.24.1

2022-10-19

### Fixes

- Fixed dates validation (2d99f6bf9c1b423065c97898229f6537274d15cd)

## 0.24.0

2022-09-22

### Features

- Added type prop (77e776f39813a4f94f9e67cde01147da50be0439)

## 0.23.0

2022-09-08

### Features

- added prop to close date picker component after a selection is made (df3ca24711e9c2474bab04a050950d045d6a5332)

## 0.22.0

2022-09-02

### Features

- preventing callback from triggering where there are erros on the selectors (0fe3620cc6b1867e33bc409214dce72aca339914)

## 0.21.0

2022-08-31

### Features

- implemented internacionalization (7582570c946dc845fd11cca8e203fb61f14b504f)

## 0.20.0

2022-08-31

### Features

- Added rules and restrictions to all components (2f89d2e2dadc7264f74ae9b9aec480e525847a97)

### Fixes

- updated documentation to cover the new coponents' props (e2ef82eb1b4cebb713f3b5663685fcc835e7acf3)
- error on const assigntment (df56de6138a5f909ce20d7b4312277370bed11c6)

## 0.19.0

2022-08-30

### Features

- Added check to prevent setting default boolean values to strings (815d18c0712cb864312fe5bcb4c3918a20787f83)

## 0.18.0

2022-08-24

### Features

- Retrieving the deselected elements to their original positions when the selectors are multiple and expected to push selected items up (80318530ce68b99d2790e07ef70224c40200625e)

## 0.17.0

2022-07-14

### Features

- showing first the selected elements on selectors that accept multiple values (0fe8a514de8bd9ce9ec773a24242da083832f67a)

## 0.16.0

2022-07-11

### Features

- removing sharedProps index from the store and managing it from timeline and list components (da4336eec081e97687b9ab8b150ab07ddd9be8b2)
- added prop to vue2 components to not trigger the store reactivity when changing values (37abe5a74c241a215bccceda2831d0814ebd8d38)

### Fixes

- added execution of change method when the child element is setting its newValue to null and has actions (08b4bd137dbd6be2f85c7d86d0862d66b16c9ee9)

## 0.15.0

2022-06-29

### Features

- added magical text field selector (2dafc6da1c1d6ba6de26c5d94a0e38ea1d31e061)
- added hour picker component (f1edc07f5cf2951fba444b45872c770c3487f6ed)

### Fixes

- added documentation for vue components (8363766dcdd3a354d2136f9b1aa3cf349548d3cf)
- setting selectos value to null when it has no items (3429043030af45925ec06aad9a98ee152dac17f6)

## 0.14.0

2022-06-28

### Features

- added setDefaultItem spec parameter and changed methods promises (01f58429c6a213befb199c39ab2185bf3fc3b1e1)
- removed unwanted condition to disable the 'now' button on the timeline component (f4f11ccb15c9c476a64175ac5e442e14598bbe88)
- added emits to notify parent component that the slider is set on the last or first element (8b75659ab9a45783f303ddfd4e115473d9652354)

### Fixes

- added conditions to set value when the selector is of type date (878dc5fd997cad9a3ccf5f8d42d94823c264e3ef)
- changes on documentation content and structure (0ce5d18d39072668389476e6f64382dc60ae1500)
- changed parameter on getValues call to use the state passed on creation (66f36244eb8d4fb30445409fc552e67b6c1a7792)
- fixed conditions to trigger the callback (d690f4a9a63cd0d2fa15c912536dff06edfd5490)

## 0.13.1

2022-06-15

### Fixes

- setting instantSelectorButtonLabel required to false (d5543020ac2799cf990837fba382485326452917)

## 0.13.0

2022-06-15

### Breaking changes

#### changed exports of static method parseUrl (fe0d71949488a2339685cb63fdd18a58943fae3f)

the method parseUrl is now independent from the store class

### Features

- added tests related to stores state (d03bc5267500c098d58a3dfa7ffae53813258a11)
- added type of newValue check (04f88155282129c51fef1b672829c8ee38c8737b)

### Fixes

- fixed setState's and setSelector's conditionals used to check if the new value is part of the selector items (178a661849ba4d58c260831d4bf2d91df2cae8fb)

## 0.12.0

2022-06-08

### Features

- Added testing for store dispatched events (785f6e24cf807d8514f900e03e8aaec2f6068787)

## 0.11.0

2022-06-08

### Features

- Added new async constructor "createStore" (f81ba5318c7ed3152153cfe8d8c9016ff0e200c1)

## 0.10.0

2022-06-03

### Features

- changes on timeline and list selectors to work agains storeElement values instead of indices (af1eb868312c22c86a6f959cb9b495ecccb4bbb0)
- added promises to control slider's advance pace (ebfbd7c7dcdd3f4cc32384b6d60a0d727b0a7270)

examples/vanilla/magical-state.js
export default async (propId, params) => {
  switch (propId) {
    case "SPATIAL_AGGREGATION":
      return await [{
        label: "country-level",
        value: "COUNTRY",
      }, {
        label: "region-level",
        value: "REGION",
      }]
    case "SPATIAL_FILTER":
      switch (params["SPATIAL_AGGREGATION"]) {
        case "COUNTRY":
          return await [{
            label: "España",
            value: "SPAIN"
          }, {
            label: "Francia",
            value: "FRANCE"
          }];
        case "REGION":
          return await [{
            label: "Galicia",
            value: "GALICIA"
          }, {
            label: "Asturias",
            value: "ASTURIAS"
          }];
      }
    case "INSTANT_FILTER":
      const date = new Date(params.DATE_FILTER);
      const today = new Date();
      if (date > today.setDate(today.getDate() - 5)) {
        return await [{
          value: [2022, 3, 31, 15, 20],
          label: "15:20",
        },
        {
          value: [2022, 3, 31, 16, 0],
          label: "16:00",
        }];
      } else {
        return await [{
          value: [2022, 3, 31, 9, 20],
          label: "09:20",
        },
        {
          value: [2022, 3, 31, 9, 30],
          label: "09:30",
        }];
      }
  }
}

